Fix VIM-987

rittelle opened the pull request:
This is my attempt to fix [VIM-987](https://youtrack.jetbrains.com/issue/VIM-987).

The IDE itself does something slightly different at the moment (see [here](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/ui/KeyStrokeAdapter.java#L85-L126)) while `JComponent#processKeyBindings(KeyEvent, boolean)` falls back on the "normal" key code when it finds no matching binding (I'd like to link the source here, but it seems to be only available as scrs.zip).

Because I don't know if I break something else with this change and I want to see if I get a response on  [IDEA-197762](https://youtrack.jetbrains.com/issue/IDEA-197762), I prefaced this with WIP.
I the meantime I am testing this change on my system.

Any feedback would be appreciated.

joker234 commented:
Since https://github.com/chenkelmann/neo2-awt-hack/issues/3, I tested your Fix and it worked like a charm. Would love to have this fix merged :)

This is the script, I used to build the plugin:
```bash
#!/bin/bash

export IDEAVIM_VERSION=0.51
export IDEAVIM_BUILDFOLDER=~/build/ideavim

rm -rf $IDEAVIM_BUILDFOLDER
git clone git@github.com:JetBrains/ideavim.git -b $IDEAVIM_VERSION $IDEAVIM_BUILDFOLDER
cd $IDEAVIM_BUILDFOLDER
git remote add rittelle git@github.com:rittelle/ideavim.git
git fetch rittelle
git cherry-pick rittelle/neo-fix
git cherry-pick e7c5a20ab0408bcb9a39ed6320a662b578946ea3 # fix gradle version
./gradlew clean buildPlugin

echo "you can now copy ${IDEAVIM_BUILDFOLDER}/build/distributions/IdeaVim-SNAPSHOT.zip to your destination system and install this file by using \"Settings | Plugins | Install plugin from disk\""
```

rittelle commented:
I updated the PR to the latest [master](https://github.com/JetBrains/ideavim/commit/a1ff484c8535131ffafc52bd215ceaca883743b6) and reverted the changes caused by the formatter.
I was using this since the creation of the PR without any problems on Linux and (rarely) Windows.

joker234 commented:
Little bit Off-Topic: One issue is still, that neo keys don't work in the built-in Terminal of Intellijs IDEs. But this is something, I can handle better than IdeaVim.

rittelle commented:
The terminal pobably uses its own input routines, so we'll have to open a bug there, too.

tynsh commented:
What is the current state of this? This is an kind of an issue for me, so without pressuring anyone: Is something blocking a merge into master?

AlexPl292 commented:
@tynsh Hi!
For now, we are just concentrated on #186 PR. After that, we will continue the review of other older pull requests. #186 is almost finished and this PR is pretty small, so I think we will review it on this or next week.

tynsh commented:
Thanks a lot! I look forward to this :heart_eyes:

rittelle commented:
@tynsh In the meantime you can compile my branch youreself and use that (that's what I'm doing since I created the PR).

I would be glad to see this merged/fixed in another way, though :) .

AlexPl292 commented:
Hi @rittelle !
Thank you for PR! I've just checked it and even it seems to work with NEO, arrow keys are broken with the regular layout. Have you checked this PR with a regular keyboard, or did I miss something?

rittelle commented:
Thanks for reviewing it :)
I thought that I did test it with a normal layout…
I just tested the arrow keys with Bone/Neo and qwertz (PyCharm Professional 2019.1 on Linux, I've set them up to be switchable using the scroll lock key) to be sure and they worked in both cases. I can't remember issues on Windows aswell, but I'll try it later today. Never tried it on mac, though.

A few additional words to the change I've made:
I used the implementation of `JComponent#processKeyBindings(KeyEvent, boolean)` (sry couldn't find sources online to link, I had the JDK source installed) and Idea‘s implementation (pre [3bf940be1dc85f12463a0055791a3a29495b5bd9](https://github.com/JetBrains/intellij-community/tree/6b42e772d638b1e54989e0fe51cd908df5267ef7), see https://youtrack.jetbrains.com/issue/IDEA-197762 ), as a ressource. I'm unsure why I didn't add the other checks from the Idea code aswell (it's been a while since I touched the code).

AlexPl292 commented:
I got broken keys right on mac :)
Thank you for [this](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/ui/KeyStrokeAdapter.java#L85-L126) link. I can see that there is some specific logic for mac and windows. Now I'm wondering if we can just replace `KeyStroke.getKeyStrokeForEvent(keyEvent)` with `KeyStrokeAdapter. getDefaultKeyStroke(keyEvent)` 🤔

mbrgm commented:
@rittelle I'm successfully using your branch to fix issues with IdeaVIM and Neo layer 4 and would love to see this PR merged.

However, I have orthogonal problems with Neo2 and Jetbrains IDEs... some keys are not working for me on the main IDE windows (e.g. L4 Escape to close the quick documentation window). Are you also facing similar problems?

joker234 commented:
Is someone still working on this PR? What is the issue where this PR is stuck?

fdietze commented:
I compiled it myself and am using it since then to get Neo2 Keyboard layout with ideavim working (Linux). It's working well, I'd like to see this being merged.

MaxGyver83 commented:
Is anybody working on this? How can I see the merge conflict directly on github.com?

Btw, even with this fix applied, the level 4 arrow keys don't work when entering a (colon) command. (Not a big problem but maybe this fix just needs to be applied at another place, too.)

aut0 commented:
Tested on CentOS7. Working arrow keys with neo and de layout. Arrow keys not working, when entering a command, as described above.

Xeranor commented:
Compiled and tested on Ubuntu 18.04 LTS. Neo "Ebene 3" works as expected.

Xeranor commented:
FYI: The current release version (0.55) contains a fix for this Problem (https://github.com/JetBrains/ideavim/commit/e863ffd9493acb79237c9ac14e248fef4078951e). That means you can go back to the official releases and this PR is done.

mbrgm commented:
I can confirm that the current version fixes this issue. Thank you very much @AlexPl292 ! This really helps! :-)

AlexPl292 commented:
Oh, I'm sorry, I absolutely forgot to log the change and notify you about the update.
It's great that everything works as expected. Thank you all for confirming it.

mbrgm commented:
One minor issue still persists... Lvl 3 Escape does not work to close e.g. Quick Documentation windows. @AlexPl292 Could this be related and easy to fix or is it a totally different matter?

AlexPl292 commented:
@mbrgm Sorry, can't answer you right now, I need to check it. Don't you know if there's an issue in YouTrack for that?

MaxGyver83 commented:
> Tested on CentOS7. Working arrow keys with neo and de layout. Arrow keys not working, when entering a command, as described above.

FYI: I have created a new issue for Neo's arrow keys not working in vim command-line mode:
https://youtrack.jetbrains.com/issue/VIM-1999

fdietze commented:
Thanks for posting. Voted on it!
